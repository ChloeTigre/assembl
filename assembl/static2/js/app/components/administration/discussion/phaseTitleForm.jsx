// @flow
import React from 'react';
import { connect } from 'react-redux';
import { OverlayTrigger, Button, FormGroup } from 'react-bootstrap';
import { I18n, Translate } from 'react-redux-i18n';
import classnames from 'classnames';
import { updatePhaseTitle, deletePhase, movePhaseUp, movePhaseDown } from '../../../actions/adminActions/timeline';
import FormControlWithLabel from '../../common/formControlWithLabel';
import { deletePhaseTooltip, upTooltip, downTooltip } from '../../common/tooltips';
import { getEntryValueForLocale } from '../../../utils/i18n';

type PhaseTitleFormProps = {
  id: string,
  title: string,
  editLocale: string,
  handleTitleChange: Function,
  handleDeleteClick: Function,
  handleUpClick: Function,
  handleDownClick: Function,
  phaseIndex: number,
  numberOfPhases: number,
  hasConflictingDates: boolean
};

export const DumbPhaseTitleForm = ({
  id,
  title,
  editLocale,
  handleTitleChange,
  handleDeleteClick,
  handleUpClick,
  handleDownClick,
  phaseIndex,
  numberOfPhases,
  hasConflictingDates
}: PhaseTitleFormProps) => {
  const phaseLabel = I18n.t('administration.timelineAdmin.phaseLabel');
  const isTitleEmpty = title.length === 0;
  const isFirst = phaseIndex === 1;
  const isLast = phaseIndex === numberOfPhases;
  const phaseSideTitleClassNames = classnames('phase-side-title', { 'phase-side-title-low': isTitleEmpty });
  const deleteButtonClassNames = classnames('admin-icons', { 'side-button-high': isTitleEmpty });
  const upButtonsClassNames = classnames('admin-icons', { 'end-items': isLast, 'side-button-high': isTitleEmpty });
  const downButtonClassNames = classnames('admin-icons', { 'end-items': isFirst, 'side-button-high': isTitleEmpty });
  return (
    <div className="phase-title-form-container">
      <div className="flex phase-title-form">
        <Translate value="administration.timelineAdmin.phase" count={phaseIndex} className={phaseSideTitleClassNames} />
        <FormGroup validationState={hasConflictingDates ? 'warning' : null}>
          <FormControlWithLabel
            key={`phase-${id}`}
            label={`${phaseLabel} ${editLocale.toUpperCase()}`}
            onChange={handleTitleChange}
            type="text"
            value={title}
            required
            className={hasConflictingDates && 'warning'}
          />
        </FormGroup>
        <div className="flex">
          {!isLast ? (
            <OverlayTrigger placement="top" overlay={downTooltip}>
              <Button onClick={handleDownClick} className={downButtonClassNames}>
                <span className="assembl-icon-down-small grey" />
              </Button>
            </OverlayTrigger>
          ) : null}
          {!isFirst ? (
            <OverlayTrigger placement="top" overlay={upTooltip}>
              <Button onClick={handleUpClick} className={upButtonsClassNames}>
                <span className="assembl-icon-up-small grey" />
              </Button>
            </OverlayTrigger>
          ) : null}
          {numberOfPhases > 1 ? (
            <OverlayTrigger placement="top" overlay={deletePhaseTooltip}>
              <Button onClick={handleDeleteClick} className={deleteButtonClassNames}>
                <span className="assembl-icon-delete grey" />
              </Button>
            </OverlayTrigger>
          ) : null}
        </div>
      </div>
      <div className="warning-label">
        {hasConflictingDates && <Translate value="administration.timelineAdmin.warningLabel" />}
      </div>
    </div>
  );
};

const mapStateToProps = (state, { id, editLocale }: PhaseTitleFormProps) => {
  const phase = state.admin.timeline.phasesById.get(id);
  return {
    title: getEntryValueForLocale(phase.get('titleEntries'), editLocale, ''),
    hasConflictingDates: phase.get('hasConflictingDates')
  };
};

const mapDispatchToProps = (dispatch, { id, editLocale }: PhaseTitleFormProps) => ({
  handleTitleChange: e => dispatch(updatePhaseTitle(id, editLocale, e.target.value)),
  handleDeleteClick: () => dispatch(deletePhase(id)),
  handleUpClick: () => dispatch(movePhaseUp(id)),
  handleDownClick: () => dispatch(movePhaseDown(id))
});

export default connect(mapStateToProps, mapDispatchToProps)(DumbPhaseTitleForm);