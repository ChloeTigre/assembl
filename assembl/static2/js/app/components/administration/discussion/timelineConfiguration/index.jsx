// @flow
import React from 'react';
import { Field } from 'react-final-form';
import { I18n, Translate } from 'react-redux-i18n';
import { connect } from 'react-redux';
import arrayMutators from 'final-form-arrays';
import { type ApolloClient, graphql, compose, withApollo } from 'react-apollo';

import TimelineQuery from '../../../../graphql/Timeline.graphql';
import TabbedContent from '../../../common/tabbedContent';
import SectionTitle from '../../../../components/administration/sectionTitle';
import AdminForm from '../../../../components/form/adminForm';
import LoadSaveReinitializeForm from '../../../../components/form/LoadSaveReinitializeForm';
import { load, postLoadFormat } from './load';
import Loader from '../../../common/loader';
import FieldArrayWithActions from '../../../form/fieldArrayWithActions';
import MultilingualTextFieldAdapter from '../../../form/multilingualTextFieldAdapter';
import DatePickerFieldAdapter from '../../../form/datePickerFieldAdapter';
import FileUploaderFieldAdapter from '../../../form/fileUploaderFieldAdapter';
import { deleteResourceTooltip, createResourceTooltip, phaseTooltip } from '../../../common/tooltips';
import { convertISO8601StringToDate } from '../../../../utils/globalFunctions';
import { validStartDate, validEndDate } from '../../landingPage/header/validate';
import manageErrorAndLoading from '../../../common/manageErrorAndLoading';

const loading = <Loader />;

type Props = {
  client: ApolloClient,
  editLocale: string,
  lang: string
};

type State = {
  phases: Array<Object>,
  startDateConflict: boolean,
  endDateConflict: boolean
};

class TimelineFields extends React.Component<Props, State> {
  static getDerivedStateFromProps(props) {
    const phases = props.phases.map(({ start, end }) => ({
      start: start ? convertISO8601StringToDate(start) : null,
      end: end ? convertISO8601StringToDate(end) : null
    }));
    return {
      phases: phases,
      startDateConflict: false,
      endDateConflict: false
    };
  }

  state = {
    phases: [],
    startDateConflict: false,
    endDateConflict: false
  };

  onStartChange = (newStart: moment$Moment, index: number) => {
    const startDateConflict = validStartDate(newStart, this.state.phases[index].end);
    const updatedPhases = [...this.state.phases];
    updatedPhases[index].start = newStart;
    this.setState({
      phases: updatedPhases,
      startDateConflict: startDateConflict
    });
  };

  onEndChange = (newEnd: moment$Moment, index: number) => {
    const endDateConflict = validEndDate(newEnd, this.state.phases[index].end);
    const updatedPhases = [...this.state.phases];
    updatedPhases[index].end = newEnd;
    this.setState({
      phases: updatedPhases,
      startDateConflict: endDateConflict
    });
  };

  render() {
    const { client, editLocale, lang } = this.props;
    return (
      <div className="discussion-admin admin-box admin-content">
        <SectionTitle
          title={I18n.t('administration.discussion.5')}
          annotation={I18n.t('administration.timelineAdmin.annotation')}
        />
        <Translate value="administration.timelineAdmin.instruction1" className="admin-instruction" />
        <LoadSaveReinitializeForm
          load={(fetchPolicy: FetchPolicy) => load(client, fetchPolicy, lang)}
          loading={loading}
          postLoadFormat={postLoadFormat}
          createMutationsPromises={() => {}}
          save={() => {}}
          validate={() => {}}
          mutators={{
            ...arrayMutators
          }}
          render={({ values, handleSubmit, pristine, submitting, form }) => (
            <AdminForm handleSubmit={handleSubmit} pristine={pristine} submitting={submitting}>
              <div className="form-container phase-creation-section">
                <FieldArrayWithActions
                  name="phases"
                  withSeparators={false}
                  titleMsgId="administration.timelineAdmin.phase"
                  tooltips={{
                    addTooltip: createResourceTooltip,
                    deleteTooltip: deleteResourceTooltip
                  }}
                  renderFields={({ name }) => (
                    <React.Fragment>
                      <Field
                        key={`phase-title-form-${name}`}
                        editLocale={editLocale}
                        name={`${name}.title`}
                        component={MultilingualTextFieldAdapter}
                        label={`${I18n.t('administration.timelineAdmin.phaseLabel')} ${editLocale.toUpperCase()}`}
                        required
                      />
                    </React.Fragment>
                  )}
                />
              </div>
              <div className="section-test">
                <div className="phase-configuration-section">
                  <Translate value="administration.timelineAdmin.instruction2" className="admin-instruction" />
                </div>
                <TabbedContent
                  type="phase"
                  divClassName="admin-content"
                  tabTitleMsgId="administration.timelineAdmin.phase"
                  tabs={values.phases.map((phase, index) => {
                    const tabTitle = `${I18n.t('administration.timelineAdmin.phase', { count: index + 1 })}`;
                    return {
                      id: phase.id,
                      title: tabTitle,
                      description: phase.description,
                      start: phase.start,
                      end: phase.end,
                      index: index
                    };
                  })}
                  renderTooltip={phaseTooltip}
                  renderBody={(phase, index) => (
                    <div className="discussion-admin admin-box admin-content">
                      <Translate value="administration.timelineAdmin.instruction1" className="admin-instruction" />
                      <Field
                        name={`phases[${phase.index}].start`}
                        component={DatePickerFieldAdapter}
                        picker={{ pickerType: I18n.t('administration.landingPage.header.startDate') }}
                        editLocale={editLocale}
                        placeHolder={I18n.t('administration.landingPage.header.timePlaceholder')}
                        showTime={false}
                        hasConflictingDate={false}
                        onDateChange={e => this.onStartChange(e, index)}
                        dateFormat="LL"
                      />
                      <Field
                        name={`phases[${phase.index}].end`}
                        component={DatePickerFieldAdapter}
                        picker={{ pickerType: I18n.t('administration.landingPage.header.endDate') }}
                        editLocale={editLocale}
                        placeHolder={I18n.t('administration.landingPage.header.timePlaceholder')}
                        showTime={false}
                        hasConflictingDate={false}
                        onDateChange={e => this.onEndChange(e, index)}
                        form={form}
                        dateFormat="LL"
                      />
                      <Translate value="administration.timelineAdmin.instruction1" className="admin-instruction" />
                      <Field
                        name={`phases[${phase.index}].image`}
                        component={FileUploaderFieldAdapter}
                        label={I18n.t('administration.landingPage.header.logoDescription')}
                      />
                      <Translate value="administration.timelineAdmin.instruction1" className="admin-instruction" />
                      <Field
                        editLocale={editLocale}
                        withAttachmentButton={false}
                        name={`phases[${phase.index}].description`}
                        component={MultilingualTextFieldAdapter}
                        label={I18n.t('administration.landingPage.header.subtitleLabel')}
                        required
                      />
                    </div>
                  )}
                />
              </div>
            </AdminForm>
          )}
        />
      </div>
    );
  }
}

const mapStateToProps = state => ({
  editLocale: state.admin.editLocale,
  lang: state.i18n.locale
});

export default compose(
  connect(mapStateToProps),
  graphql(TimelineQuery, {
    options: ({ editLocale }) => ({
      variables: {
        lang: editLocale
      }
    })
  }),
  manageErrorAndLoading({ displayLoader: true }),
  withApollo
)(TimelineFields);