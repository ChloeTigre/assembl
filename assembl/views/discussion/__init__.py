
def find_discussion_from_slug(request, slug, session=None):
    from assembl.models import Discussion, OldSlug
    db = session or Discussion.default_db
    old_slugs = {o.slug: o for o in db.query(OldSlug)}
    if slug in old_slugs:
        found = False
        while not found:
            redirection_slug = old_slugs[slug].redirection_slug
            if redirection_slug in old_slugs:
                del old_slugs[redirection_slug]
            else:
                found = True
                discussion = db.query(Discussion).filter_by(slug=redirection_slug).first()
                return discussion
    else:
        return db.query(Discussion).filter_by(slug=slug).first()

def includeme(config):
    config.include('.views')
