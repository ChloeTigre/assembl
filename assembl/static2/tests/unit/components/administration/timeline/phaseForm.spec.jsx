import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import moment from 'moment';
import { DumbPhaseForm } from '../../../../../js/app/components/administration/discussion/phaseForm';

describe('PhaseForm component', () => {
  const handleIdentifierChangeSpy = jest.fn(() => {});
  const handleStartDateChangeSpy = jest.fn(() => {});
  const handleEndDateChangeSpy = jest.fn(() => {});
  const commonProps = {
    phaseId: '1234',
    phaseNumber: 1,
    handleIdentifierChange: handleIdentifierChangeSpy,
    handleStartDateChange: handleStartDateChangeSpy,
    handleEndDateChange: handleEndDateChangeSpy,
    identifier: 'survey',
    start: moment('2014-12-31T09:00:00+00:00'),
    end: moment('2015-12-31T09:00:00+00:00'),
    locale: 'fr'
  };
  it('should render a form to update the dates and module of a new phase', () => {
    const props = {
      ...commonProps,
      isNew: true
    };

    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbPhaseForm {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
  it('should render a form to update an old phase, hence with a disabled menu for module', () => {
    const props = {
      ...commonProps,
      isNew: false
    };
    const shallowRenderer = new ShallowRenderer();
    shallowRenderer.render(<DumbPhaseForm {...props} />);
    const result = shallowRenderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});