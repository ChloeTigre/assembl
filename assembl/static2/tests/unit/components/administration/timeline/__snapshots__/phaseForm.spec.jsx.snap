// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhaseForm component should render a form to update an old phase, hence with a disabled menu for module 1`] = `
<div
  className="phase-form"
>
  <div
    className="date-picker-field"
  >
    <div
      className="date-picker-type"
    >
      <Translate
        value="search.datefilter.from"
      />
    </div>
    <label
      className="datepicker-label"
      htmlFor="start-datepicker"
    >
      <DatePicker
        allowSameDay={false}
        className=""
        dateFormat="LLL"
        dateFormatCalendar="MMMM YYYY"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        id="start-datepicker"
        locale="fr"
        monthsShown={1}
        onBlur={[Function]}
        onChange={[MockFunction]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Phase 1's start date"
        preventOpenOnFocus={false}
        selected={"2014-12-31T09:00:00.000Z"}
        shouldCloseOnSelect={true}
        showTimeSelect={true}
        timeCaption="Time"
        timeFormat="HH:mm"
        timeIntervals={30}
        withPortal={false}
      />
      <div
        className="icon-schedule-container"
      >
        <span
          className="assembl-icon-schedule grey"
        />
      </div>
    </label>
  </div>
  <div
    className="date-picker-field"
  >
    <div
      className="date-picker-type"
    >
      <Translate
        value="search.datefilter.to"
      />
    </div>
    <label
      className="datepicker-label"
      htmlFor="end-datepicker"
    >
      <DatePicker
        allowSameDay={false}
        className=""
        dateFormat="LLL"
        dateFormatCalendar="MMMM YYYY"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        id="end-datepicker"
        locale="fr"
        monthsShown={1}
        onBlur={[Function]}
        onChange={[MockFunction]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Phase 1's end date"
        preventOpenOnFocus={false}
        selected={"2015-12-31T09:00:00.000Z"}
        shouldCloseOnSelect={true}
        showTimeSelect={true}
        timeCaption="Time"
        timeFormat="HH:mm"
        timeIntervals={30}
        withPortal={false}
      />
      <div
        className="icon-schedule-container"
      >
        <span
          className="assembl-icon-schedule grey"
        />
      </div>
    </label>
  </div>
  <div
    className="module-selection-text"
  >
    <Translate
      value="administration.timelineAdmin.phaseModule"
    />
  </div>
  <div
    className="margin-m"
  >
    <OverlayTrigger
      defaultOverlayShown={false}
      overlay={
        <Tooltip
          bsClass="tooltip"
          id="moduleSelectionTooltip"
          placement="right"
        >
          <Translate
            value="administration.timelineAdmin.moduleSelectionTooltip"
          />
        </Tooltip>
      }
      placement="top"
      trigger={
        Array [
          "hover",
          "focus",
        ]
      }
    >
      <div>
        <SplitButton
          className="admin-dropdown"
          disabled={true}
          id="dropdown-1234"
          onSelect={[MockFunction]}
          title="Survey module"
        >
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="survey"
            header={false}
          >
            Survey module
          </MenuItem>
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="thread"
            header={false}
          >
            Thread module
          </MenuItem>
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="multiColumns"
            header={false}
          >
            Multi-columns module
          </MenuItem>
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="voteSession"
            header={false}
          >
            Vote module
          </MenuItem>
          <MenuItem
            bsClass="dropdown"
            disabled={false}
            divider={false}
            eventKey="brightMirror"
            header={false}
          >
            Bright Mirror module
          </MenuItem>
        </SplitButton>
      </div>
    </OverlayTrigger>
  </div>
  <div
    className="text-xs configure-module-text"
  >
    <Translate
      value="administration.timelineAdmin.configureModule"
    />
    <Link
      className="configure-module-link"
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/\${slug}/administration/survey?phaseId=1234&section=1"
    >
      <Translate
        count={1}
        value="administration.timelineAdmin.configureModuleLink"
      />
    </Link>
  </div>
</div>
`;

exports[`PhaseForm component should render a form to update the dates and module of a new phase 1`] = `
<div
  className="phase-form"
>
  <div
    className="date-picker-field"
  >
    <div
      className="date-picker-type"
    >
      <Translate
        value="search.datefilter.from"
      />
    </div>
    <label
      className="datepicker-label"
      htmlFor="start-datepicker"
    >
      <DatePicker
        allowSameDay={false}
        className=""
        dateFormat="LLL"
        dateFormatCalendar="MMMM YYYY"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        id="start-datepicker"
        locale="fr"
        monthsShown={1}
        onBlur={[Function]}
        onChange={[MockFunction]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Phase 1's start date"
        preventOpenOnFocus={false}
        selected={"2014-12-31T09:00:00.000Z"}
        shouldCloseOnSelect={true}
        showTimeSelect={true}
        timeCaption="Time"
        timeFormat="HH:mm"
        timeIntervals={30}
        withPortal={false}
      />
      <div
        className="icon-schedule-container"
      >
        <span
          className="assembl-icon-schedule grey"
        />
      </div>
    </label>
  </div>
  <div
    className="date-picker-field"
  >
    <div
      className="date-picker-type"
    >
      <Translate
        value="search.datefilter.to"
      />
    </div>
    <label
      className="datepicker-label"
      htmlFor="end-datepicker"
    >
      <DatePicker
        allowSameDay={false}
        className=""
        dateFormat="LLL"
        dateFormatCalendar="MMMM YYYY"
        disabled={false}
        disabledKeyboardNavigation={false}
        dropdownMode="scroll"
        id="end-datepicker"
        locale="fr"
        monthsShown={1}
        onBlur={[Function]}
        onChange={[MockFunction]}
        onClickOutside={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onMonthChange={[Function]}
        onSelect={[Function]}
        onYearChange={[Function]}
        placeholderText="Phase 1's end date"
        preventOpenOnFocus={false}
        selected={"2015-12-31T09:00:00.000Z"}
        shouldCloseOnSelect={true}
        showTimeSelect={true}
        timeCaption="Time"
        timeFormat="HH:mm"
        timeIntervals={30}
        withPortal={false}
      />
      <div
        className="icon-schedule-container"
      >
        <span
          className="assembl-icon-schedule grey"
        />
      </div>
    </label>
  </div>
  <div
    className="module-selection-text"
  >
    <Translate
      value="administration.timelineAdmin.phaseModule"
    />
  </div>
  <div
    className="margin-m"
  >
    <SplitButton
      className="admin-dropdown"
      disabled={false}
      id="dropdown-1234"
      onSelect={[MockFunction]}
      title="Survey module"
    >
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        eventKey="survey"
        header={false}
      >
        Survey module
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        eventKey="thread"
        header={false}
      >
        Thread module
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        eventKey="multiColumns"
        header={false}
      >
        Multi-columns module
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        eventKey="voteSession"
        header={false}
      >
        Vote module
      </MenuItem>
      <MenuItem
        bsClass="dropdown"
        disabled={false}
        divider={false}
        eventKey="brightMirror"
        header={false}
      >
        Bright Mirror module
      </MenuItem>
    </SplitButton>
  </div>
  <div
    className="text-xs configure-module-text"
  >
    <Translate
      value="administration.timelineAdmin.configureModule"
    />
    <Link
      className="configure-module-link"
      onlyActiveOnIndex={false}
      style={Object {}}
      to="/\${slug}/administration/survey?phaseId=1234&section=1"
    >
      <Translate
        count={1}
        value="administration.timelineAdmin.configureModuleLink"
      />
    </Link>
  </div>
</div>
`;
